python/tchannel/tornado/hyperbahn.py

```diff
@@ -59,33 +64,43 @@ def advertise(tchannel, service, routers):
         tchannel.peers.get(router)
 
     @tornado.gen.coroutine
-    def _register():
-        response = yield tchannel.request(service='hyperbahn').send(
-            arg1='ad',  # advertise
-            arg2='',
-            arg3=json.dumps({
-                'services': [
-                    {
-                        'serviceName': service,
-                        'cost': 0,
-                    }
-                ]
-            }),
-            headers={'as': 'json'},
-        )
-
-        if response.code not in ErrorCode:
-            # re-register every ``DEFAULT_TTL`` seconds
-            tornado.ioloop.IOLoop.current().call_later(
-                delay=DEFAULT_TTL,
-                callback=_register,
+    def _register(attempt_counter=0):
+
+        try:
+            response = yield tchannel.request(service='hyperbahn').send(
+                arg1='ad',  # advertise
+                arg2='',
+                arg3=json.dumps({
+                    'services': [
+                        {
+                            'serviceName': service,
+                            'cost': 0,
+                        }
+                    ]
+                }),
+                headers={'as': 'json'},
+                attempt_times=1,
             )
-            raise tornado.gen.Return(response)
+        except (ProtocolError, TimeoutError) as e:
+            attempt_counter += 1
+            log.error('Failed to register with Hyperbahn: %s', e)
         else:
-            log.error('Failed to register with Hyperbahn: %s', response)
-            raise NotImplementedError  # TODO figure out behavior here
+            if response.code != StatusCode.ok:
+                attempt_counter += 1
+                log.error('Failed to register with Hyperbahn: %s', response)
+            else:
+                attempt_counter = 0
+
+        delay_time = random.uniform(
+            0, min(DEFAULT_MAX_DELAY, DEFAULT_EXPO_BASE ** attempt_counter)
+        )
 
-    return _register()
+        attempt_counter = min(attempt_counter, DEFAULT_MAX_ATTEMPT)
+        tornado.ioloop.IOLoop.current().call_later(
+            delay=delay_time,
+            callback=lambda: _register(attempt_counter),
+        )
```

Comment:
If you `raise Return(response)` here, `_register()` *will* return the response of the first call. That's how `return _register()` was able to make the result of the first call available as a future.